src/python/pants/backend/python/rules/python_fmt.py

```diff
@@ -0,0 +1,137 @@
+from pants.engine.isolated_process import ExecuteProcessRequest, ExecuteProcessResult
+from pants.engine.legacy.structs import (
+  PantsPluginAdaptor,
+  PythonAppAdaptor,
+  PythonBinaryAdaptor,
+  PythonTargetAdaptor,
+  PythonTestsAdaptor,
+)
+from pants.engine.rules import UnionRule, optionable_rule, rule
+from pants.engine.selectors import Get
+from pants.rules.core.fmt import FmtResult, FmtTarget
+from pants.util.objects import datatype
+
+# Note: this is a workaround until https://github.com/pantsbuild/pants/issues/8343 is addressed
+# We have to write this type which basically represents a union of all various kinds of targets
+# containing python files so we can have one single type used as an input in the run_black rule.
+class FormattablePythonTarget(datatype(['target'])):
+  pass
+
+
+@rule
+def run_black(
+  wrapped_target: FormattablePythonTarget,
+  black: Black,
+  python_setup: PythonSetup,
+  subprocess_encoding_environment: SubprocessEncodingEnvironment,
+  ) -> FmtResult:
+  config_path = black.get_options().config
+  config_snapshot = yield Get(Snapshot, PathGlobs, PathGlobs(include=(config_path,)))
+
+  resolved_requirements_pex = yield Get(
+    RequirementsPex, RequirementsPexRequest(
+      output_filename="black.pex",
+      requirements=tuple(black.get_requirement_specs()),
+      interpreter_constraints=tuple(black.default_interpreter_constraints),
+      entry_point=black.get_entry_point(),
+    )
+  )
+  target = wrapped_target.target
+  sources_digest = target.sources.snapshot.directory_digest
+
+  all_input_digests = [
+    sources_digest,
+    resolved_requirements_pex.directory_digest,
+    config_snapshot.directory_digest,
+  ]
+  merged_input_files = yield Get(
+    Digest,
+    DirectoriesToMerge,
+    DirectoriesToMerge(directories=tuple(all_input_digests)),
+  )
+
+  # The exclude option from Black only works on recursive invocations,
+  # so call black with the directories in which the files are present
+  # and passing the full file names with the include option
+  dirs = set()
+  for filename in target.sources.snapshot.files:
+    dirs.add(os.path.dirname(filename))
+  pex_args= tuple(sorted(dirs))
+  if config_path:
+    pex_args += ("--config", config_path)
+  if target.sources.snapshot.files:
+    pex_args += ("--include", "|".join(re.escape(file) for file in target.sources.snapshot.files))
+
+  request = resolved_requirements_pex.create_execute_request(
+    python_setup=python_setup,
+    subprocess_encoding_environment=subprocess_encoding_environment,
+    pex_path="./black.pex",
+    pex_args=pex_args,
+    input_files=merged_input_files,
+    output_files=target.sources.snapshot.files,
+    description=f'Run Black for {target.address.reference()}',
+  )
+
+  result = yield Get(ExecuteProcessResult, ExecuteProcessRequest, request)
+
+  yield FmtResult(
+    digest=result.output_directory_digest,
+    stdout=result.stdout.decode(),
+    stderr=result.stderr.decode(),
+  )
+
+
+@rule
```

Comment:
Please add a TODO on these definitions pointing to the ticket(s) you opened about unions.